File: src/app/useCases/trafficSource/createTrafficSourceUseCase.ts
import { TrafficSource } from "../../../domain/entities/trafficSource";
import { UserGatewayDTO } from "../../../domain/gateways/user";
import { TrafficSourceRepository } from "../../../domain/repositories/trafficSource";
import { HttpAdapter } from "../../../infra/adapters/httpAdapter";

type InputProps = {
  name: string;
  trafficDomain: string;
};

class CreateTrafficSourceUseCase {
  constructor(
    private trafficSourceRepository: TrafficSourceRepository,
    private userGateway: UserGatewayDTO
  ) {}

  async execute(input: InputProps, token: string) {
    const { name, trafficDomain } = input;

    const [existsDomain, user] = await Promise.all([
      this.trafficSourceRepository.findByDomain(trafficDomain),
      this.userGateway.findUnique(token),
    ]);

    if (existsDomain) {
      throw HttpAdapter.conflict("Traffic domain already exists");
    }

    const trafficSource = TrafficSource.create({
      name,
      userId: user.id,
      trafficDomain,
    });

    await this.trafficSourceRepository.createTrafficSource(trafficSource);

    return trafficSource.toJson(user.utc);
  }
}

export { CreateTrafficSourceUseCase };

File: internal/app/useCases/trafficSource/createTrafficSourceUseCase.go
package trafficSourceUseCases

import (
	"github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/entities"
	domainErrors "github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/errors"
	"github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/providers"
	"github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/repositories"
)

type CreateTrafficSourceUseCase struct {
	trafficSourceRepository repositories.TrafficSourceRepository
	dateGenerator           providers.DateGenerator
	uuidGenerator           providers.UuidGenerator
}

type CreateTrafficSourceInput struct {
	Name          string
	UserId        string
	TrafficDomain string
}

func NewCreateTrafficSourceUseCase(
	trafficSourceRepository repositories.TrafficSourceRepository,
	dateGenerator providers.DateGenerator,
	uuidGenerator providers.UuidGenerator,
) *CreateTrafficSourceUseCase {
	return &CreateTrafficSourceUseCase{
		trafficSourceRepository: trafficSourceRepository,
		dateGenerator:           dateGenerator,
		uuidGenerator:           uuidGenerator,
	}
}

func (t *CreateTrafficSourceUseCase) Handle(input *CreateTrafficSourceInput) (*entities.TrafficSource, error) {
	trafficSourceByDomain, _ := t.trafficSourceRepository.FindByDomain(input.UserId, input.TrafficDomain)

	if trafficSourceByDomain != nil {
		return nil, domainErrors.AlreadyExists("There is already a traffic source for this domain.")
	}

	trafficSource := entities.NewTrafficSource(
		&entities.NewTrafficSourceInput{Name: input.Name, UserId: input.UserId, TrafficDomain: input.TrafficDomain},
		t.uuidGenerator,
		t.dateGenerator,
	)

	_, createRepositoryErr := t.trafficSourceRepository.Create(trafficSource)

	if createRepositoryErr != nil {
		return nil, createRepositoryErr
	}

	return trafficSource, nil
}
